consensus/sync/sync_test.go

```
@@ -0,0 +1,178 @@
+	stateMachine.EXPECT().ProcessProposal(gomock.Any()).Return(
+		[]types.Action[starknet.Value, starknet.Hash, starknet.Address]{},
+	)
+
+	proposalCh := make(chan starknet.Proposal)
+	prevoteCh := make(chan starknet.Prevote)
+	precommitCh := make(chan starknet.Precommit)
+	p2p := newMockP2P(proposalCh, prevoteCh, precommitCh)
+
+	driver := driver.New(
+		utils.NewNopZapLogger(),
+		newTendermintDB(t),
+		stateMachine,
+		newMockBlockchain(t),
+		p2p,
+		newMockProposer(),
+		mockTimeoutFn,
+	)
+
+	go func() {
+		require.NoError(t, driver.Run(ctx))
+	}()
+
+	mockInCh := make(chan sync.BlockBody)
+	mockP2PSyncService := newMockP2PSyncService(mockInCh)
+	proposalStore := proposal.ProposalStore[starknet.Hash]{}
+
+	consensusSyncService := consensusSync.New(&mockP2PSyncService, proposalCh, precommitCh, getPrecommits, toValue, &proposalStore)
+
+	block0 := getCommittedBlock()
+	block0Hash := block0.Block.Hash
+	valueHash := toValue(block0Hash).Hash()
+	go func() {
+		mockP2PSyncService.recieveBlockOverP2P(block0)
+	}()
+
+	go func() {
+		time.Sleep(2 * time.Second)
+		cancel()
+	}()
+	consensusSyncService.Run(ctx)                     // Driver should trigger stopSyncCh and shut this service down
+	require.NotEmpty(t, proposalStore.Get(valueHash)) // Ensure the Driver sees the correct proposal
+}
+
+func TestShutdownOnError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx, cancel := context.WithCancel(t.Context())
+
+	proposalCh := make(chan starknet.Proposal)
+	precommitCh := make(chan starknet.Precommit)
+
+	mockInCh := make(chan sync.BlockBody)
+	mockP2PSyncService := newMockP2PSyncService(mockInCh)
+	proposalStore := proposal.ProposalStore[starknet.Hash]{}
+
+	consensusSyncService := consensusSync.New(&mockP2PSyncService, proposalCh, precommitCh, getPrecommits, toValue, &proposalStore)
+	cancel()
+	consensusSyncService.Run(ctx)
+
+	mockP2PSyncService.shouldTriggerErr(true)
+	consensusSyncService.Run(t.Context())
```

`Run` should return an error and here we should check each time for the kind of error we are expecting. 

Additionally, I think this error(s) should be logged somewhere, in the context of production code.